frame/election-provider-multi-phase/src/mock.rs

```diff
@@ -615,7 +615,24 @@ impl ExtBuilder {
 	}
 
 	pub fn build_and_execute(self, test: impl FnOnce() -> ()) {
-		self.build().execute_with(test)
+		sp_tracing::try_init_simple();
+
+		let mut ext = self.build();
+		ext.execute_with(test);
+
+		// run `try-runtime` post conditions logic at the end of each test, even if
+		// `try-runtime` feature is disabled.
+		#[cfg(not(feature = "try-runtime"))]
+		ext.execute_with(|| {
+			assert_ok!(MultiPhase::do_try_state());
+		});
+
+		#[cfg(feature = "try-runtime")]
+		ext.execute_with(|| {
+			assert_ok!(<MultiPhase as frame_support::traits::Hooks<u64>>::try_state(
```

Comment:
Do we need this? Why not just call `do_try_state()` irrespective of `try-runtime` feature?

Reply:
My goal was to also verify in the tests that the `try_state` hook was being properly implemented and called as if if was the executive. I do agree that this makes the mock more verbose.

I'm inclined to leave this as is for now and then work on https://github.com/paritytech/substrate/issues/13463 as a follow-up to remove all this complexity and verbosity, wdyt?

Reply:
In practice, there is really no difference. At the end of the test, the same checks under `MultiPhase::do_try_state()` will be called in both cases.

However, the idea here is that if the `try-runtime` feat is enabled, the `try_state` hook should be called directly, instead of the `do_try_state()`. The try_state hook cannot be called if try runtime is not enable -- however, now that I'm thinking about it, we could have:

```rust
  #[cfg(any(feature = "try-runtime", test))]
  fn try_state(_n: T::BlockNumber) -> Result<(), &'static str> {
    Self::do_try_state()
  }
```

and simplify the mock code by calling the hook in both tests and try-runtime.

Reply:
> However, the idea here is that if the try-runtime feat is enabled, the try_state hook should be called directly, instead of the do_try_state().

I still don't understand the benefit of this idea

Reply:
I kinda see what @gpestana is doing here: if you run `cargo test` **without** `--features try-runtime` we want try_state hooks to still exeucte. 

I think for the sake of simplicity, it is fine to assume that we only want to run `try_state` when `--features try-runtime` is enabled. Instead of patching too many things in the code, we should make sure the CI runs tests with this feature. 

Also, if you want to do that, the simpler and more comprehensive thing to do would be to call `<AllPallets as TryState>::try_state()`. 

Sadly there is no way to make `fn try_state` as it is be `#[cfg(any(test, feature = "try-state"))]` as `test` is not a compiler condition that is propagated across crates. I might be wrong though, and to check, you can try and make all the `fn try_state` stuff in `frame_support` be `#[cfg(any(test, feature = "try-state"))]` and see if that would make it available here in the test context or not.

Reply:
Yep, annotating `try_state` with `#[cfg(any(test, feature = "try-state"))]` doesn't work.

I'll keep it simple and assume that if devs want to run try_state, then they'll run with the correct feature.